product/src/main/java/com/yas/product/service/ProductService.java

```diff
@@ -111,73 +111,98 @@ public ProductService(ProductRepository productRepository,
-                        newProductImages.add(ProductImage.builder()
-                                .imageId(imageId).product(variantInDb).build());
-                    }
-                });
-            }
+        Product mainProduct = Product.builder()
+            .name(productPostVm.name())
+            .thumbnailMediaId(productPostVm.thumbnailMediaId())
+            .slug(productPostVm.slug())
+            .description(productPostVm.shortDescription())
+            .shortDescription(productPostVm.description())
+            .specification(productPostVm.specification())
+            .sku(productPostVm.sku())
+            .gtin(productPostVm.gtin())
+            .price(productPostVm.price())
+            .isAllowedToOrder(productPostVm.isAllowedToOrder())
+            .isPublished(productPostVm.isPublished())
+            .isFeatured(productPostVm.isFeatured())
+            .isVisibleIndividually(productPostVm.isVisibleIndividually())
+            .stockTrackingEnabled(productPostVm.stockTrackingEnabled())
+            .metaTitle(productPostVm.metaTitle())
+            .metaKeyword(productPostVm.metaKeyword())
+            .metaDescription(productPostVm.description())
+            .hasOptions(CollectionUtils.isNotEmpty(productPostVm.variations())
+                && CollectionUtils.isNotEmpty(productPostVm.productOptionValues()))
+            .productCategories(List.of())
+            .taxClassId(productPostVm.taxClassId())
+            .build();
+
+        setProductBrand(productPostVm.brandId(), mainProduct);
+
+        List<ProductCategory> productCategories = setProductCategories(productPostVm.categoryIds(), mainProduct);
+
+        List<ProductImage> productImages = setProductImages(productPostVm.productImageIds(), mainProduct);
+
+        Product savedMainProduct = productRepository.saveAndFlush(mainProduct);
+        productImageRepository.saveAllAndFlush(productImages);
+        productCategoryRepository.saveAllAndFlush(productCategories);
+
+        // create product relations
+        createProductRelations(productPostVm, savedMainProduct);
+
+        // create product variations, options and combinations
+        if (CollectionUtils.isEmpty(productPostVm.variations())
+            || CollectionUtils.isEmpty(productPostVm.productOptionValues())) {
+            return ProductGetDetailVm.fromModel(savedMainProduct);
         }
-    }
 
-    public ProductListGetVm getProductsWithFilter(int pageNo, int pageSize, String productName, String brandName) {
-        Pageable pageable = PageRequest.of(pageNo, pageSize);
-        Page<Product> productPage;
+        // create product variations and their images
+        List<Product> savedVariations = createProductVariations(productPostVm, mainProduct, Collections.emptyList());
+        if (CollectionUtils.isEmpty(savedVariations)) {
+            throw new InternalServerErrorException(Constants.ErrorCode.FAILED_TO_SAVE_VARIATIONS);
```

Comment:
In which case might this error occur? Please add a log to provide more details and help identify the potential root cause of the error.
In case, the exception should be thrown inside `createProductVariations` method.